File: tools-java-gen/src/main/kotlin/com/afoxer/javagen/StructGenerator.kt
package com.afoxer.javagen

import com.squareup.javapoet.FieldSpec
import com.squareup.javapoet.JavaFile
import com.squareup.javapoet.TypeSpec
import java.io.Serializable
import javax.lang.model.element.Modifier

class StructGenerator(val desc: StructDesc, val pkg: String) {
    fun generate(): JavaFile.Builder {
        val builder = TypeSpec.classBuilder(this.desc.name)
                .addSuperinterface(Serializable::class.java)
                .addModifiers(Modifier.PUBLIC)

        val fields = desc.fields
        for (field in fields) {
            val fieldBuilder = FieldSpec.builder(mapType(field.ty, field.sub_ty, false), field.name, Modifier.PUBLIC)
            builder.addField(fieldBuilder.build())
        }
        return JavaFile.builder(this.pkg, builder.build())
    }
}

File: toolchain/swift-gen/Sources/SwiftGenCore/StructGenerator.swift
//
//  File.swift
//  SwiftGenCore
//
//  Created by wangxin.sidney on 2019/2/17.
//

import Foundation
import SwiftPoet

class StructGenerator {
    private let structDesc: StructDesc
    private let libModName: String
    
    init(desc: StructDesc, libModName: String) {
        self.structDesc = desc
        self.libModName = libModName
    }
    
    public func generate() -> StructSpecBuilder {
        let classBuilder = StructSpecBuilder.init(name: "\(self.structDesc.name)")
//             .add(import: self.libModName)
            .add(modifier: .Public)
            .add(superType: TypeName.init(keyword: "Codable"))
        
        for field in self.structDesc.fields {
            classBuilder.add(field: FieldSpec.builder(for: field.name, type: field.ty.toTypeName()).add(modifier: Modifier.Public).build())
        }
        
        return classBuilder
    }
}
